src/documents/models.py

```diff
@@ -483,8 +483,14 @@ def from_path(cls, path):
           "<title>.<suffix>"
         """
 
+        filename = os.path.basename(path)
+        for (pattern, repl) in settings.FILENAME_PARSE_TRANSFORMS:
+            (filename, count) = pattern.subn(repl, filename)
+            if count:
```

Comment:
If I'm reading this right, if the for loop exits with `count` never being greater than `0`, then `filename` will just be whatever is last in the list of parse transforms.  It'd be safer to write it something like this:

```python
filename = os.path.basename(path)
transformed_filename = ""

for (pattern, repl) in settings.FILENAME_PARSE_TRANSFORMS:
    (transformed_filename, count) = pattern.subn(repl, filename)
    if count:
        break

for regex in cls.REGEXES.values():
    m = regex.match(transformed_filename or filename)
```

Reply:
> If I'm reading this right, if the for loop exits with `count` never being greater than `0`, then `filename` will just be whatever is last in the list of parse transforms.

Nope, it won't. `re.subn` is like search/replace. It returns the modified input string, and the number of matches (= replace operations) that happened. If there are no matches, no modifications happened, count is zero and the returned string equals the string passed into the function. It never returns the transformation pattern (`repl`) it tried to apply.

Example at the python cli to demonstrate:

```
>>> import re
>>> re.compile('more').subn('less', 'papermore')
('paperless', 1)
>>> re.compile('less').subn('less', 'papermore')
('papermore', 0)
```

So there are two possible outcomes in each interation of the `for` loop:

  1. There was a match and `re.subn` altered the filename. In this case `count` is *not* zero and therefore we break out of the loop, as we only want to apply at most one transformation.

  2. There was no match, `filename` stays unaltered (as `re.subn` returned the unaltered string) and we go to the next iteration. If this happened to be the last iteration, the `for` loop ends naturally and `filename` still equals the value it had before entering the loop (which is `os.path.basename(path)`)

See also: https://docs.python.org/3/library/re.html#re.subn

Reply:
Excellent.  Thank you for walking me through that.  As my mother likes to say, you learn something new every day :-)  This all looks good.  Sorry to complicate things.